### PR TITLE
Add missing types required to support aquilon

### PR DESCRIPTION
**Pull request by jrha**
We've been carrying these in an aquilon branch in the repository up until now, the recent restructuring of the schema forced me to look at integrating this again.

### CODE REVIEW

File: quattor/types/system.pan [hunk 1]
```diff
@@ -28,6 +35,105 @@ type structure_oldname = {
...
+    "filesystem-layout" ? string with if_exists("archetype/filesystem-layouts/" + SELF) != ""
+    "archlist"      ? string[] # e.g. fs sysname list for model,
+                               # "x86_64.linux.2.6.glibc.2.3", "amd64.linux.2.4.glibc.2.3", ...
+};
+
+type structure_event = {
+    "timeout"       ? long
+    "actions"       ? string[] with is_action_list(SELF)
+};
+type structure_events = {
```
Comment (stdweird) [inline quattor/types/system.pan]: newline

File: quattor/types/system.pan [hunk 2]
```diff
@@ -28,6 +35,105 @@ type structure_oldname = {
     "date" : string
 };
 
+type structure_archetype = {
```
Comment (stdweird) [inline quattor/types/system.pan]: can we move these in a `quattor/types/aquilon` template and include it here? i don't like this one growing again

File: quattor/types/system.pan [hunk 3]
```diff
@@ -61,4 +167,18 @@ type structure_system = {
     "state"         ? string with match (SELF,
         '^(production|standby|test|development|onloan)$') # "production|out-of-production|test|development|onloan"
     "vo"            ? structure_vo{}
+    # Aquilon properties
```
Comment (stdweird) [inline quattor/types/system.pan]: make an `aquilon_system_type` and include it here

File: quattor/types/system.pan [hunk 4]
```diff
@@ -40,6 +147,22 @@ include 'quattor/types/aii';
...
+    "security"      ? structure_security
+    "users"         ? nlist
+    "eon_ids"       ? long[]
+    "eon_id_maps"   ? nlist
+    "owner_eon_id"  ? long
+    "provides"      ? nlist
+    "services"      ? structure_services{}
+    "personality"   ? structure_personality
+    "host_environment" ? string
+    "features"      ? nlist
```
Comment (jrha) [inline quattor/types/system.pan]: @gombasg I can't find this used in our hosts, was it moved to metadata and this got left behind?

File: quattor/types/system.pan [hunk 5]
```diff
@@ -40,6 +147,22 @@ include 'quattor/types/aii';
...
+}
```
Comment (stdweird) [inline quattor/types/system.pan]: missing `;`, causing failure in https://jenkins1.ugent.be/job/NCM_Components-pull_requests/705/org.quattor.cfg.module$opennebula/console

**Comment by stdweird** [comment]
test this please

**Comment by stdweird** [comment]
test this please

File: quattor/types/system.pan [hunk 6]
```diff
@@ -17,6 +17,13 @@ type structure_export = {
 type structure_cluster = {
     "name" : string
     "type" ? string
+    "resources"     ? structure_resources
```
Comment (stdweird) [inline quattor/types/system.pan]: `structure_resources` used before defined

File: quattor/types/system.pan [hunk 7]
```diff
@@ -61,4 +184,5 @@ type structure_system = {
     "state"         ? string with match (SELF,
         '^(production|standby|test|development|onloan)$') # "production|out-of-production|test|development|onloan"
     "vo"            ? structure_vo{}
+    include structure_system_aquilon;
```
Comment (stdweird) [inline quattor/types/system.pan]: no `;`

**Comment by stdweird** [comment]
test this please

**Comment by stdweird** [comment]
@jrha you have the doubtful honour to test new template tests for this repo.
except for some newly introduced bugs, you will also have to cleanup the existing ones, esp in the components.

run the following until no errors occur.

```bash
rm -Rf build_temp
mkdir -p build_temp/monitoring/lemon
echo "object template test;" > build_temp/test.pan
echo "declaration template monitoring/lemon/schema; type structure_monitoring = {};" > build_temp/monitoring/lemon/schema.pan
find pan -type f ! -regex '^./build_temp/.*' -name *.pan | xargs sed -n "s/^declaration[ ]\+template[ ]\+\(.*\);/include '\1';/p" >> build_temp/test.pan
find . -type f ! -regex '^./pan/.*' ! -regex '^./build_temp/.*' -name *.pan | xargs sed -n "s/^declaration[ ]\+template[ ]\+\(.*\);/include '\1';/p" >> build_temp/test.pan
find . -type f ! -regex '^./build_temp/.*' -name *.pan | xargs sed -n "s/^\(unique[ ]\+\)\?template[ ]\+\(.*\);/include '\2';/p" >> build_temp/test.pan
panc --output-dir build_temp --include-path .:build_temp build_temp/test.pan
```

**Comment by jrha** [comment]
:speak_no_evil: I accept the "honour".

File: quattor/types/system.pan (hunk 4 above)
Comment (gombasg) [inline quattor/types/system.pan]: /system/features is used to pass parameters to featues. So parameter "bar" for feature "foo" will end up like `"/system/features/foo/bar" = ...`

**Comment by gombasg** [comment]
Btw. there are ideas to ship the schema of broker-generated data with the broker itself. Right now, if we want to export more data from the broker, that means templates need to be updated first, and the scheduling of changes can be problematic sometimes.

But letting the broker export the schema would work only if there was a proper way of extending/subclassing/inheriting from existing types. Otherwise, we'd end up having to use `extensible` everywhere, which would not be ideal either.

File: quattor/types/aquilon.pan [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+
+type structure_security = {
+    "class"         : string with if_exists("archetype/security/" + SELF) != ""
+    "svcwhitelist"  ? list
+};
+ 
+type structure_system_aquilon = {
+    "advertise_status" ? boolean
+    "archetype"     ? structure_archetype
+    "build"         ? string with match (SELF, 'build|blind|ready|failed|install')
```
Comment (stdweird) [inline quattor/types/aquilon.pan]: stricter regexp `^(build|blind|ready|failed|install)$`

File: quattor/types/aquilon.pan [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+@documentation{
+    aquilon-related structures
+}
+
+type structure_fsdata = {
+    "mountpoint"        : string
+    "mountopts"         ? string
+    "type"              : string
+    "mount"             ? boolean
+    "pass"              ? long
```
Comment (stdweird) [inline quattor/types/aquilon.pan]: can be `long(..)`?

File: quattor/types/aquilon.pan [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+};
+
+type structure_provides = {
+    "instance"      : string
+    "clients"       ? string[]
+};
+
+type structure_security = {
+    "class"         : string with if_exists("archetype/security/" + SELF) != ""
+    "svcwhitelist"  ? list
```
Comment (stdweird) [inline quattor/types/aquilon.pan]: list of what? (if not type, maybe example?)

File: quattor/types/aquilon.pan [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+};
+ 
+type structure_system_aquilon = {
+    "advertise_status" ? boolean
+    "archetype"     ? structure_archetype
+    "build"         ? string with match (SELF, 'build|blind|ready|failed|install')
+    "security"      ? structure_security
+    "users"         ? nlist
```
Comment (stdweird) [inline quattor/types/aquilon.pan]: what is the key/value ? maybe an example?

File: quattor/types/metadata.pan [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+    "type" : string
+};
+
+type structure_template = {
+    "branch" : structure_branch
+};
+
+type structure_metadata = {
+    "template" : structure_template
+    "features" ? list
```
Comment (stdweird) [inline quattor/types/metadata.pan]: list of what? examples?

**Comment by stdweird** [comment]
@jrha more restrictive types could be better, but afaik, this comes out of the broker?
@gombasg is ok for you that you ship the relevant schema in dedicated types like `quattor/types/aquilon` as part of the broker. that way, we can just copy them here in the proper location without modifications.

File: quattor/types/aquilon.pan (hunk 10 above)
Comment (jrha) [inline quattor/types/aquilon.pan]: I have no idea... @gombasg?

**Comment by jrha** [comment]
@stdweird are you happy with this now?

**Comment by stdweird** [comment]
yes, tests fail due to something else. i'll merge this.

**Pull request merged**